File: src/de/saschahlusiak/frupic/services/FetchFrupicJob.java
package de.saschahlusiak.frupic.services;

import de.saschahlusiak.frupic.model.Frupic;
import de.saschahlusiak.frupic.model.FrupicFactory;
import de.saschahlusiak.frupic.model.FrupicFactory.OnFetchProgress;

public class FetchFrupicJob extends Job implements OnFetchProgress {
	Frupic frupic;
	FrupicFactory factory;
	
	public FetchFrupicJob(Frupic frupic, FrupicFactory factory) {
		this.frupic = frupic;
		this.factory = factory;
	}

	@Override
	public JobState run() {
		try {			
			factory.fetchFrupicImage(frupic, false, this);
		}
		catch (InterruptedException e) {
			return JobState.JOB_CANCELLED;
		}
		
		return JobState.JOB_SUCCESS;
	}
	
	public Frupic getFrupic() {
		return frupic;
	}
	
	@Override
	public void OnProgress(int read, int length) {
		synchronized (jobListener) {
			for (OnJobListener l: jobListener) {
				l.OnJobProgress(this, read, length);
			}
		}
	}
}

File: app/src/main/java/de/saschahlusiak/frupic/services/FetchFrupicJob.kt
package de.saschahlusiak.frupic.services

import android.util.Log
import androidx.annotation.WorkerThread
import de.saschahlusiak.frupic.app.FrupicManager
import de.saschahlusiak.frupic.model.Frupic
import kotlinx.coroutines.runBlocking

@Deprecated("")
class FetchFrupicJob constructor(val frupic: Frupic, private val manager: FrupicManager) : Job() {
    public override fun run(): JobState {
        Log.d(Companion.tag, "fetching #" + frupic.id)
        if (!doFetch()) {
            Log.d(Companion.tag, "failed #" + frupic.id)
            return JobState.JOB_FAILED
        }
        Log.d(Companion.tag, "fetched #" + frupic.id)
        return JobState.JOB_SUCCESS
    }

    /**
     * @return true on success, false otherwise
     */
    @WorkerThread
    fun doFetch(): Boolean {
        return try {
            runBlocking {
                manager.download(frupic) { _, copied, max ->
                    synchronized(jobListener) {
                        for (l in jobListener) {
                            l.OnJobProgress(this@FetchFrupicJob, copied, max)
                        }
                    }
                }
                true
            }
        } catch (e: Exception) {
            e.printStackTrace()
            false
        }
    }

    companion object {
        private val tag = FetchFrupicJob::class.java.simpleName
    }
}